No predeclared identifiers, `Core` is a keyword (#4864)
carbon-language/carbon-lang: docs/project/principles/namespace_cleanliness.md

-->

<!-- toc -->

## Table of contents

-   [Background](#background)
-   [Principle](#principle)
-   [Applications of this principle](#applications-of-this-principle)
-   [Exceptions](#exceptions)
-   [Alternatives considered](#alternatives-considered)

<!-- tocstop -->

## Background

Names and entities in a program can come from multiple sources -- from a local
declaration, from an import, from the standard library, or from the prelude.
Names can be imported from Carbon code or imported or derived from code written
in another language such as C++ or an interface description language such as
that of Protobuf, MIDL, or CORBA. Names can be selected for use in a program
that language designers later decide they want to use as keywords. And in order
to use a library, it is sometimes necessary to redeclare the same name that that
library chose.

This puts a lot of pressure on the language to support a free choice of naming
for entities. Different languages make different choices in this space:

-   Many languages have a set of keywords that are not usable as identifiers,
    with no workaround. If this set collides with a name needed by user code,

>     the user is left to solve this problem, often by picking an uglified
>     identifier (`klass` or `class_`), which sometimes conflicts with the general

Reviewer: ```suggestion
    the user is left to solve this problem, often by rewriting the
    identifier in some way (`klass` or `class_`), which sometimes conflicts with the general
```

    naming convention used by the code. And conversely, suboptimal choices are
    made for new language keywords to avoid causing problems for existing code.
-   C and C++ reserve a family of identifiers, such as those beginning with an
    underscore and a capital letter. However, it's not clear which audiences the
    reserved identifiers are for, and this leads to collisions between standard
    library vendors and compiler authors, as well as between implementation
    extensions and language extensions.
    -   MSVC provides a `__identifier(keyword)` extension that allows using a

>         keyword as an identifier.

Reviewer: Maybe also mention GCC and Clang equivalent extensions?

Author: Done.

-   Python reserves some identifiers but still allows them to be freely
    overwritten (such as `bool`) and reserves some identifiers but rejects
    assignment to them (such as `True`).
-   Rust provides a raw identifier syntax to allow most identifiers with
    reserved meaning to be used by a program, but
    [not all](https://internals.rust-lang.org/t/raw-identifiers-dont-work-for-all-identifiers/9094):
    `self`, `Self`, `super`, `extern`, and `crate` cannot be used as raw
    identifiers. Rust also predeclares a large number of library names in every
    file, but allows them to be shadowed by user declarations with the same
    name.
-   Swift provides a raw identifier syntax using backticks: `` `class` ``, and
    is
    [considering](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0451-escaped-identifiers.md)
    extending this to allow arbitrary non-word-shaped character sequences
    between the `` ` ``s.

Carbon provides raw identifier syntax, for example `r#for`, to allow using
keywords as identifiers. Carbon also has strict shadowing rules that mean that
predeclared identifiers that are _not_ keywords are difficult or impossible to

> redeclare and use in inner scopes.

Reviewer: Maybe link to the relevant docs when describing what Carbon provides?

Author: Done.

## Principle

In Carbon, the language does not encroach on the developer's namespace. There
are no predeclared or reserved identifiers. In cases where the language gives
special meaning to a word or to word-shaped syntax such as `i32`, that special
meaning can always be undone with raw identifier syntax, `r#`.

Conversely, when adding language keywords, we will not select an inferior
keyword merely to avoid the risk of breaking existing programs. We will still
take into account how often it is desirable to use the word as an identifier,
including in domain-specific contexts, because that is a factor in whether it

> would make a good keyword.

Reviewer: Maybe a brief mention that we'll also manage rolling out these kinds of changes to enable easy migration of conflicts to use `r#` syntax?

Author: Done.
